### PR TITLE
initial implementation of destroy and publish in digital objects API

### PR DESCRIPTION
**Pull request by barmintor**

### CODE REVIEW

File: lib/hyacinth/adapters/publication_adapter/hyacinth2.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          update_doi(digital_object, location) if point_doi_to_this_publish_target && location
+          true
+        rescue StandardError => e
+          [false, e.message]
+        end
+
+        # @return [success, errors] [boolean, Array<String>] success will be true if
+        #         the unpublish was successful, or false otherwise. errors is an array
+        #         that will contain error messages if the unpublish failed.
+        def unpublish(publish_target, digital_object)
```
Comment (cgalarza) [inline lib/hyacinth/adapters/publication_adapter/hyacinth2.rb]: I think this method is missing updating the DOI entry if the digital object is unpublished from the "primary" publish target.

File: lib/hyacinth/adapters/publication_adapter/hyacinth2.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+          digital_object_pid = digital_object_pids(digital_object).first
+          return [false, "Never preserved to Fedora3"] unless digital_object_pid
+          connection = Faraday.new(publish_target.publish_url)
+          connection.token_auth(publish_target.api_key)
+          connection.delete(digital_object_pid)
+          true
```
Comment (cgalarza) [inline lib/hyacinth/adapters/publication_adapter/hyacinth2.rb]: Should this also be an array to follow the pattern?

I have to be honest I don't love sending the success and error message back in an array like this, but right now I can't think of a better solution. Maybe using a hash? I'll keep thinking about it.

File: spec/requests/api/v1/digital_objects_spec.rb [hunk 3]
```diff
@@ -118,18 +120,71 @@
...
+  describe 'POST /api/v1/digital_objects/:id/publish' do
+    include_examples 'requires user to have correct permissions' do
+      let(:request) do
+        post "/api/v1/digital_objects/#{authorized_object.uid}/publish"
+      end
+    end
+
+    context "logged in" do
+      let(:publication_adapter) { instance_double(Hyacinth::Adapters::PublicationAdapter::Abstract) }
+      let(:configured_publication_adapter) { Hyacinth.config.publication_adapter }
```
Comment (cgalarza) [inline spec/requests/api/v1/digital_objects_spec.rb]: Can use `let!` here to automatically run this.

File: spec/requests/api/v1/digital_objects_spec.rb [hunk 4]
```diff
@@ -118,18 +120,71 @@
...
+      end
+      it 'unpublishes the object on delete' do
+        expect(publication_adapter).to receive(:publish).with(authorized_publish_target, digital_object_matcher, anything).and_return([true, []])
+        post "/api/v1/digital_objects/#{authorized_object.uid}/publish"
+      end
+      it "return json entity" do
+        expect(publication_adapter).to receive(:publish).with(authorized_publish_target, digital_object_matcher, anything).and_return([true, []])
+        post "/api/v1/digital_objects/#{authorized_object.uid}/publish"
+        new_object = JSON.parse(response.body)
+        expect(new_object['projects'].first).to include('string_key' => authorized_project.string_key)
```
Comment (cgalarza) [inline spec/requests/api/v1/digital_objects_spec.rb]: Should this be checking that the publish target on the object was set correctly?

File: spec/requests/api/v1/digital_objects_spec.rb (hunk 3 above)
Comment (barmintor) [reply spec/requests/api/v1/digital_objects_spec.rb]: that's awesome & new to me

File: lib/hyacinth/adapters/publication_adapter/hyacinth2.rb (hunk 2 above)
Comment (barmintor) [reply lib/hyacinth/adapters/publication_adapter/hyacinth2.rb]: Yes, to all of this.

File: app/models/concerns/digital_object_concerns/destroy_behavior.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+          self.digital_object_record.optimistic_lock_token = self.mint_optimistic_lock_token
+
+          # remove metadata storage
+          self.remove_from_metadata_storage
+          all_targets = self.projects.map do |project|
+            project.publish_targets.map(&:string_key)
+          end.flatten.uniq
+          self.set_pending_publish_entries({ 'publish_to' => [], 'unpublish_from' => all_targets })
+          # if everything worked, destroy digital object record
+          self.digital_object_record.destroy!
```
[the remark this answers is not in the document]
Comment (barmintor) [reply app/models/concerns/digital_object_concerns/destroy_behavior.rb]: there's not really an indexing service yet... I'll need to TODO it

File: app/models/concerns/digital_object_concerns/destroy_behavior.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
+module DigitalObjectConcerns
+  module DestroyBehavior
+    extend ActiveSupport::Concern
+
+    # This method is like the other #save method, but it raises an error if the save fails.
```
[the remark this answers is not in the document]
Comment (barmintor) [reply app/models/concerns/digital_object_concerns/destroy_behavior.rb]: Yep.

File: app/models/concerns/digital_object_concerns/destroy_behavior.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (barmintor) [reply app/models/concerns/digital_object_concerns/destroy_behavior.rb]: You know I'll just add the stub call to this and the save behavior.

File: app/models/concerns/digital_object_concerns/publish_behavior.rb [hunk 7]
```diff
@@ -82,6 +65,31 @@ def publish(opts = {})
       self.errors.blank?
     end
 
+    def publish_to(publish_target, current_publish_entries, opts = {})
```
Comment (cgalarza) [inline app/models/concerns/digital_object_concerns/publish_behavior.rb]: Giant 👍

**Pull request merged**